src/async_reader/mod.rs

```diff
@@ -473,6 +478,198 @@ impl<T: for<'a> ZarrReadAsync<'a> + Clone + Unpin + Send + 'static>
     }
 }
 
+//***********************************************************
+// implementation of an async stream that doesn't block when decompressing chunks. for now, that doesn't
+// to help at all, I need to revisit this at some point, won't be used for now. Also, for now filters are
+// not supported, i needed to add a "wrapper" for the mask stream, that packagaes the stream and the returned
+// chunk into a future, to avoid static lifetime issues.
+//************************************************************
+
+type InterleavedResults<T> = (
+    Result<StoreReadResults<T>, tokio::task::JoinError>,
+    Result<ZarrResult<RecordBatch>, tokio::task::JoinError>,
+);
+enum ZarrStreamStateNonBlocking<T: ZarrStream> {
+    Init,
+    Reading(BoxFuture<'static, StoreReadResults<T>>),
+    _Processing(ZarrRecordBatchReader<ZarrInMemoryChunkContainer>),
+    Interleaving(Option<ZarrRecordBatchReader<ZarrInMemoryChunkContainer>>),
+    ProcessingInterleaved(BoxFuture<'static, InterleavedResults<T>>),
+    Done,
+    Error,
+}
+
+pub struct ZarrRecordBatchStreamNonBlocking<'a, T: ZarrStream> {
+    meta: ZarrStoreMetadata,
+    filter: Option<ZarrChunkFilter>,
+    store_wrapper: Option<ZarrStoreWrapper<T>>,
+    state: ZarrStreamStateNonBlocking<T>,
+
+    // this is an optional record batch stream that will provide masks
+    // to apply to the data.
+    _mask_stream: Option<BoxStream<'a, ZarrResult<RecordBatch>>>,
+    _mask: Option<BooleanArray>,
+}
+
+impl<'a, T: ZarrStream> ZarrRecordBatchStreamNonBlocking<'a, T> {
+    fn new(meta: ZarrStoreMetadata, filter: Option<ZarrChunkFilter>, store: T) -> Self {
+        Self {
+            meta,
+            filter,
+            store_wrapper: Some(ZarrStoreWrapper::new(store)),
+            state: ZarrStreamStateNonBlocking::Init,
+            _mask_stream: None,
+            _mask: None,
+        }
+    }
+}
+
+impl<'a, T> Stream for ZarrRecordBatchStreamNonBlocking<'a, T>
+where
+    T: ZarrStream + Unpin + Send + 'static,
+{
+    type Item = ZarrResult<RecordBatch>;
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        loop {
+            match &mut self.state {
+                ZarrStreamStateNonBlocking::Init => {
+                    let wrapper = self.store_wrapper.take().expect(LOST_STORE_ERR);
+                    let fut = wrapper.get_next().boxed();
+                    self.state = ZarrStreamStateNonBlocking::Reading(fut);
+                }
+                ZarrStreamStateNonBlocking::Reading(f) => {
+                    let (wrapper, chunk) = ready!(f.poll_unpin(cx));
+                    self.store_wrapper = Some(wrapper);
+
+                    // if store returns none, it's the end and it's time to return
+                    if chunk.is_none() {
```

Comment:
e.g.

```rust
let chunk = if let Some(chunk) = chunk {
    chunk // or maybe do more here
} else {
  self.state = ZarrStreamStateNonBlocking::Done;
  return Poll::Ready(None);
}

Reply:
yeah you're right, I've been following a few patterns that are not really rust-idiomatic, I've been learning rust on-and-off for the last 2 years or so and I sometimes get sloppy, like these `.is_none` followed by an `unwrap` call. I'll do a quick check tonight and clean it up a bit.